### PR TITLE
ieee-36: POST: /api/event/teams/join/<team_code>/

### PR DESCRIPTION
**Pull request by arshinmar**
## Overview

- Resolves #305 
- Added a JoinTeamAPIView
- Added corresponding tests
- Added a "create event team" functionality with a customizable user number

## Unit Tests Created

- Check the JoinTeamTestCase in event/test_api.py

## Steps to QA

- @grahamhoyes you know best...

### CODE REVIEW

File: hackathon_site/event/api_urls.py [hunk 1]
```diff
@@ -7,5 +7,10 @@
 urlpatterns = [
     path("users/user/", api_views.CurrentUserAPIView.as_view(), name="current-user"),
     path("teams/team/", views.CurrentTeamAPIView.as_view(), name="current-team"),
```
Comment (grahamhoyes) [inline hackathon_site/event/api_urls.py:9]: I just noticed that this view is in `views` instead of `api_views`, whoops. Will make a ticket to fix that

File: hackathon_site/event/api_views.py [hunk 2]
```diff
@@ -1,10 +1,25 @@
 from django.db import transaction
 from django.db.models import Q
 
+from event.models import User, Team
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: We separate our imports into 3 sections:
- Standard library imports
- 3rd party imports (django, rest_framework, etc)
- Local imports (event, hardware, etc)

Can you rearrange these so that everything is together properly?

File: hackathon_site/event/api_views.py [hunk 3]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
         response_data = response_serializer.data
         return Response(data=response_data, status=status.HTTP_201_CREATED,)
+
+
+class JoinTeamView(generics.GenericAPIView, mixins.RetrieveModelMixin):
+    permission_classes = [UserHasProfile]
+    serializer_class = TeamSerializer
+
+    def clean(self):
+        if not is_registration_open():
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: This is for the event, not registration.

From the ticket description, the actual check we need to have:

If the team they are currently on has any orders, HTTP 400 and don’t allow them to leave, with a message in the response.

File: hackathon_site/event/api_views.py [hunk 4]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
         response_serializer = TeamSerializer(profile.team)
         response_data = response_serializer.data
         return Response(data=response_data, status=status.HTTP_201_CREATED,)
+
+
+class JoinTeamView(generics.GenericAPIView, mixins.RetrieveModelMixin):
+    permission_classes = [UserHasProfile]
+    serializer_class = TeamSerializer
+
+    def clean(self):
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: I don't think that `clean` is a method used by DRF views, it's a django form method

File: hackathon_site/event/api_views.py [hunk 5]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+        if not is_registration_open():
+            raise forms.ValidationError(
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: Is your IDE properly configured, and did you test this? `forms` is never imported.

Nor should it be - look at how the `LeaveTeamView` raises a `ValidationError` (which is a DRF error, different from `forms.ValidationError`)

File: hackathon_site/event/api_views.py [hunk 6]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+
+    @transaction.atomic
+    def post(self, request, team_code, *args, **kwargs):
+
+        profile = request.user.profile
+        current_team = profile.team
+
+        try:
+            team = Team.objects.get(team_code=team_code)
+        except:
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: It's bad practice to have a generic `except`, since it may swallow errors that we haven't thought about (your IDE should be warning you of this).

Instead, do `except Team.DoesNotExist`, which is the error that will be thrown if the team doesn't exist.

File: hackathon_site/event/api_views.py [hunk 7]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+            raise ValidationError(
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: When something is not found, in most cases (including this case) we want to return a HTTP 404, not 400. 

You could do that by raising an `Http404` error from `django.http`, or constructing a drf `Response` with a 404 status code.

My recommendation though: the generic API view has a method called `get_object()` which will automatically look up the object for you and 404 if necessary. By defult it uses the `pk` argument on a view, but you can set it to filter on team code by adding `lookup_field = "team_code"` to your view class.

[Here's](https://github.com/encode/django-rest-framework/blob/master/rest_framework/generics.py#L75) the source code of `get_object()` if you're curious.

File: hackathon_site/event/api_views.py (hunk 2 above)
Comment (grahamhoyes) [reply hackathon_site/event/api_views.py]: Looks like you have a lot of duplicate imports in here too

File: hackathon_site/event/api_views.py [hunk 8]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+                code="registration_closed",
+            )
+
+        return super().clean()
+
+    @transaction.atomic
+    def post(self, request, team_code, *args, **kwargs):
+
+        profile = request.user.profile
+        current_team = profile.team
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py:81]: Unused, that's concerning

File: hackathon_site/event/api_views.py [hunk 9]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+
+        try:
+            team = Team.objects.get(team_code=team_code)
+        except:
+            raise ValidationError(
+                {"detail": "Team does not exist!"}, code=status.HTTP_400_BAD_REQUEST
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: We don't use "!" in our error responses

File: hackathon_site/event/api_views.py [hunk 10]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+            )
+
+        if not team.profiles.exists():
+            raise ValidationError(
+                {"detail": "Cannot join a fully empty team!"},
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: Why not? Imo this isn't a check we need since there's no harm in joining an empty team. It shouldn't happen anyway since we clean up teams when they get empty.

File: hackathon_site/event/api_views.py (hunk 3 above)
Comment (grahamhoyes) [reply hackathon_site/event/api_views.py]: There's also code for that as part of the `LeaveTeamView` already

Comment (grahamhoyes) [reply hackathon_site/event/api_views.py]: Ok, all of the checks for this are part of the post method already. So why is this method still here?

File: hackathon_site/event/api_views.py [hunk 11]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+            ~Q(order__status="Cancelled"), Q(order__team=team),
+        )
+
+        if active_orders.exists():
+            raise ValidationError(
+                {"detail": "Cannot join a team with already processed orders"},
+                code=status.HTTP_400_BAD_REQUEST,
+            )
+
+        print(type(profile.team))
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: remove

File: hackathon_site/event/api_views.py [hunk 12]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+
+        if not team.profiles.exists():
+            raise ValidationError(
+                {"detail": "Cannot join a fully empty team!"},
+                code=status.HTTP_400_BAD_REQUEST,
+            )
+        if team.profiles.count() >= Team.MAX_MEMBERS:
+            raise ValidationError({"detail": "Team is full"})
+
+        # Raise 400 if team has active orders
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: We don't want to let them leave a team if the team they are on has active orders (which is what the ticket says) - I don't see any reason to not let them join a team with orders, they just won't be able to leave it once they join

File: hackathon_site/event/api_views.py [hunk 13]
```diff
@@ -66,3 +81,57 @@ def post(self, request, *args, **kwargs):
...
+        )
+
+        if active_orders.exists():
+            raise ValidationError(
+                {"detail": "Cannot join a team with already processed orders"},
+                code=status.HTTP_400_BAD_REQUEST,
+            )
+
+        print(type(profile.team))
+        profile.team = team
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py:97]: You haven't handled cleaning up their old team at all

File: hackathon_site/event/test_api.py [hunk 14]
```diff
@@ -52,42 +52,77 @@ def test_user_has_profile(self):
         self.assertEqual(response.json(), serializer.data)
 
 
-class CurrentTeamTestCase(SetupUserMixin, APITestCase):
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]: ... so you deleted an entire view's tests? Why?

File: hackathon_site/event/test_api.py [hunk 15]
```diff
@@ -52,42 +52,77 @@ def test_user_has_profile(self):
...
+        self.view_name = "api:event:join-team"
 
-        self.view = reverse("api:event:current-team")
+    def _build_view(self, team_code):
+        return reverse(self.view_name, kwargs={"team_code": team_code})
+
+    def test_invalid_key(self):
+        self._login()
+        response = self.client.post(self._build_view("56ABD"))
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]: Should be 404

File: hackathon_site/event/test_api.py [hunk 16]
```diff
@@ -52,42 +52,77 @@ def test_user_has_profile(self):
...
+
+    def test_join_full_team(self):
+        self._login()
+        team = self._make_full_event_team(self_users=False)
+        print(team.team_code)
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]: remove

File: hackathon_site/event/test_api.py [hunk 17]
```diff
@@ -52,42 +52,77 @@ def test_user_has_profile(self):
...
+        response = self.client.post(self._build_view(team.team_code))
+        self.assertEqual(str(response.content), 'b\'{"detail":"Team is full"}\'')
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]: Look at some of our other API tests. You can do `data = response.json()` to get a dict, then compare that dict rather than an ugly string like this

File: hackathon_site/event/test_api.py [hunk 18]
```diff
@@ -52,42 +52,77 @@ def test_user_has_profile(self):
...
-    def test_user_has_profile(self):
+    def check_cannot_join(self):
+        old_team = self.profile.team
+        response = self.client.post(self._build_view(self.team.team_code))
+        self.user.refresh_from_db()
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(old_team.pk, self.user.profile.team.pk)
+        self.assertEqual(Team.objects.filter(team_code=old_team.team_code).count(), 1)
+
+    def test_leave_with_order(self):
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]: this isn't the leave team view

File: hackathon_site/hackathon_site/tests.py [hunk 19]
```diff
@@ -25,6 +26,20 @@ def setUp(self):
     def _login(self):
         self.client.login(username=self.user.username, password=self.password)
 
+    @staticmethod
+    def _apply_to_profile_as_user(user, team=None):
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]: Why `_apply_to_profile`? What does this have to do with applying? Perhaps `_make_profile` would be a better name

File: hackathon_site/hackathon_site/tests.py [hunk 20]
```diff
@@ -62,73 +77,134 @@ def _get_random_email():
...
-    def _make_full_registration_team(self, team=None, self_users=True):
+    def _make_full_registration_team(self, team=None, self_users=True, num_users=4):
         if team is None:
             team = RegistrationTeam.objects.create()
 
         if self_users:
+
             user1 = self.user
+            cumulative_users = [user1]
+            if num_users > 1:
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]: You can come up with a better way to write this using a for loop.

I don't care about using the painter's names either (unless we use them elsewhere in tests), you can just generate random values for all the fields.

File: hackathon_site/hackathon_site/tests.py [hunk 21]
```diff
@@ -62,73 +77,134 @@ def _get_random_email():
...
+            email1 = self._get_random_email()
+            user1 = User.objects.create_user(
+                username=email1,
+                password="foobar123",
+                email=email1,
+                first_name="John1",
+                last_name="Doe1",
+            )
+            cumulative_users = [user1]
+            if num_users > 1:
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]: Same, if you want to make this configurable, this one is really easy to make using a for loop

File: hackathon_site/event/test_api.py (hunk 15 above)
Comment (arshinmar) [reply hackathon_site/event/test_api.py]: for some reason, I can't seem to resolve this issue. Will push code soon. The views code seems to make sense

File: hackathon_site/event/test_api.py (hunk 14 above)
Comment (arshinmar) [reply hackathon_site/event/test_api.py]: sorry about that

File: hackathon_site/event/api_urls.py (hunk 1 above)
Comment (arshinmar) [reply hackathon_site/event/api_urls.py:9]: ok, sounds good

File: hackathon_site/event/api_views.py (hunk 2 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: resolved in new code

File: hackathon_site/event/api_views.py (hunk 3 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: removed in new code, part of edward's view, so thought it would be worth initially taking

File: hackathon_site/event/api_views.py (hunk 4 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: removed

File: hackathon_site/event/api_views.py (hunk 5 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: removed

File: hackathon_site/event/api_views.py (hunk 6 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: resolved in new code

File: hackathon_site/event/api_views.py (hunk 7 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: added get_object functionality in new code

File: hackathon_site/event/api_views.py (hunk 8 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py:81]: utilised in new code

File: hackathon_site/event/api_views.py (hunk 9 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: fixed

File: hackathon_site/event/api_views.py (hunk 10 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: removed

File: hackathon_site/event/api_views.py (hunk 11 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: done

File: hackathon_site/event/api_views.py (hunk 12 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py]: now, members cannot join another team if current team has active orders, as per expectation/description of JIRA issue

File: hackathon_site/event/api_views.py (hunk 13 above)
Comment (arshinmar) [reply hackathon_site/event/api_views.py:97]: now, members cannot join another team if current team has active orders, as per expectation/description of JIRA issue

if old team only had one member (themselves), the team is deleted.

File: hackathon_site/event/test_api.py (hunk 15 above)
Comment (arshinmar) [reply hackathon_site/event/test_api.py]: resolved now

File: hackathon_site/event/test_api.py (hunk 16 above)
Comment (arshinmar) [reply hackathon_site/event/test_api.py]: done

File: hackathon_site/event/test_api.py (hunk 17 above)
Comment (arshinmar) [reply hackathon_site/event/test_api.py]: done

File: hackathon_site/event/test_api.py (hunk 18 above)
Comment (arshinmar) [reply hackathon_site/event/test_api.py]: changed up the hardware consideration as per above comments.

File: hackathon_site/hackathon_site/tests.py (hunk 19 above)
Comment (arshinmar) [reply hackathon_site/hackathon_site/tests.py]: changed

File: hackathon_site/hackathon_site/tests.py (hunk 20 above)
Comment (arshinmar) [reply hackathon_site/hackathon_site/tests.py]: fixed

File: hackathon_site/hackathon_site/tests.py (hunk 21 above)
Comment (arshinmar) [reply hackathon_site/hackathon_site/tests.py]: done with a for loop

File: hackathon_site/event/api_views.py [hunk 22]
```diff
@@ -1,3 +1,4 @@
+from event.models import Team
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: Already imported as `EventTeam` below

File: hackathon_site/event/api_views.py [hunk 23]
```diff
@@ -66,3 +67,41 @@ def post(self, request, *args, **kwargs):
         response_serializer = TeamSerializer(profile.team)
         response_data = response_serializer.data
         return Response(data=response_data, status=status.HTTP_201_CREATED,)
+
+
+class JoinTeamView(generics.GenericAPIView, mixins.RetrieveModelMixin):
+    permission_classes = [UserHasProfile]
+    serializer_class = TeamSerializer
+    lookup_field = "team_code"
+    queryset = Team
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]:
```suggestion
    queryset = EventTeam.objects.all()
```

File: hackathon_site/event/api_views.py [hunk 24]
```diff
@@ -66,3 +67,41 @@ def post(self, request, *args, **kwargs):
...
+
+    @transaction.atomic
+    def post(self, request, team_code, *args, **kwargs):
+
+        profile = request.user.profile
+        current_team = profile.team
+
+        self.kwargs["lookup_url_kwarg"] = team_code
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: `lookup_url_kwarg` is the URL keyword argument that gets used in filtering. It basically does:

```python
team = EventTeam.objects.filter(**{self.lookup_field: self.kwargs[self.lookup_url_kwarg]})
```

Which we want to be

```python
EventTeam.objects.filter(team_code=team_code)
```

`self.lookup_url_kwarg` will default to `self.lookup_field`, so you don't need to do this. This is setting it on self.kwargs anyway, which is doesn't do anything.

File: hackathon_site/event/api_views.py [hunk 25]
```diff
@@ -66,3 +67,41 @@ def post(self, request, *args, **kwargs):
...
+        team = self.get_object()
+
+        if team.profiles.count() >= Team.MAX_MEMBERS:
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]:
```suggestion
        if team.profiles.count() >= EventTeam.MAX_MEMBERS:
```

File: hackathon_site/event/api_views.py [hunk 26]
```diff
@@ -66,3 +67,41 @@ def post(self, request, *args, **kwargs):
...
+
+        profile.team = team
+        profile.save()
+
+        if not current_team.profiles.exists():
+            current_team.delete()
+
+        response_serializer = TeamSerializer(profile.team)
+        response_data = response_serializer.data
+        return Response(data=response_data, status=status.HTTP_201_CREATED,)
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: Response status here can just be 200 OK

File: hackathon_site/event/test_api.py [hunk 27]
```diff
@@ -90,6 +90,102 @@ def test_user_has_profile(self):
         self.assertEqual(response.json(), serializer.data)
 
 
+class JoinTeamTestCase(SetupUserMixin, APITestCase):
+    def setUp(self):
+        super().setUp()
+        self.group = Group.objects.create(name="Test Users")
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]: Unnecessary

File: hackathon_site/hackathon_site/tests.py [hunk 28]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
...
+                    username=random_email,
+                    password="foobar123",
+                    email=random_email,
+                    first_name="John{}".format(user_number),
+                    last_name="Doe{}".format(user_number),
+                )
+                self._apply_as_user(new_user, team)
+        return team
 
+    def _make_full_event_team(self, team=None, self_users=True, num_users=4):
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]: Nit: Since this isn't making a full team, can just call it `_make_event_team`. Same for the registration one above, if you use pycharm to rename the method (right click > refactor > rename) it should update everywhere.

File: hackathon_site/hackathon_site/tests.py [hunk 29]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
             email = f"{uuid[:10]}@{uuid[10:20]}.com"
         return email
 
-    def _make_full_registration_team(self, team=None, self_users=True):
+    def create_user_set(self):
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]:
```suggestion
    def _create_user_set(self):
```

File: hackathon_site/hackathon_site/tests.py [hunk 30]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
...
-                last_name="Carmichael",
-            )
-            user4 = self.user4 = User.objects.create_user(
-                username="lawren@harris.com",
-                password="wxyz7890",
-                email="lawren@harris.com",
-                first_name="Lawren",
-                last_name="Harris",
-            )
+        if self_users == True:
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]:
```suggestion
        if self_users:
```

File: hackathon_site/hackathon_site/tests.py [hunk 31]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
...
-                email=email4,
-                first_name="John4",
-                last_name="Doe4",
-            )
-
-        self._apply_as_user(user1, team)
-        self._apply_as_user(user2, team)
-        self._apply_as_user(user3, team)
-        self._apply_as_user(user4, team)
+            for user_number in range(1, num_users + 1, 1):
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]:
```suggestion
            for user_number in range(1, num_users + 1):
```

File: hackathon_site/hackathon_site/tests.py [hunk 32]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
...
+                random_email = self._get_random_email()
+                new_user = User.objects.create_user(
+                    username=random_email,
+                    password="foobar123",
+                    email=random_email,
+                    first_name="John{}".format(user_number),
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]: Slightly more condensed:

```suggestion
                    first_name=f"John{user_number}",
```

Same with other format strings below

File: hackathon_site/hackathon_site/tests.py [hunk 33]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
...
+                    last_name="Doe{}".format(user_number),
+                )
+                self._apply_as_user(new_user, team)
+        return team
 
+    def _make_full_event_team(self, team=None, self_users=True, num_users=4):
+        if team is None:
+            team = EventTeam.objects.create()
+        if self_users == True:
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]:
```suggestion
        if self_users:
```

File: hackathon_site/hackathon_site/tests.py [hunk 34]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
...
+            for user in self.create_user_set():
+                self._apply_as_user(user, team)
+        else:
+            for user_number in range(1, num_users + 1, 1):
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]:
```suggestion
            for user_number in range(1, num_users + 1):
```

File: hackathon_site/hackathon_site/tests.py [hunk 35]
```diff
@@ -62,74 +76,67 @@ def _get_random_email():
...
+                random_email = self._get_random_email()
+                new_user = User.objects.create_user(
+                    username=random_email,
+                    password="foobar123",
+                    email=random_email,
+                    first_name="John{}".format(user_number),
```
Comment (grahamhoyes) [inline hackathon_site/hackathon_site/tests.py]:
```suggestion
                    first_name=f"John{user_number}",
```

File: hackathon_site/event/test_api.py [hunk 36]
```diff
@@ -90,6 +90,102 @@ def test_user_has_profile(self):
...
+    def _build_view(self, team_code):
+        return reverse(self.view_name, kwargs={"team_code": team_code})
+
+    def test_join_and_delete(self):
+        """
+        When a member wants to join a team, if their current team only includes
+        them, their current team is deleted.
+        """
+        self._login()
+        self.client.post(self._build_view(self.team.team_code))
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]: Instead of using a post request to put the user on a team the first time, can you just set it up directly from a model query? Helps keep this test isolated in what it is testing (that leaving an empty team deletes it)

File: hackathon_site/event/test_api.py [hunk 37]
```diff
@@ -90,6 +90,102 @@ def test_user_has_profile(self):
...
+        sample_team = self._make_full_event_team(self_users=False, num_users=2)
+        response = self.client.post(self._build_view(sample_team.team_code))
+        self.user.refresh_from_db()
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(response.data["id"], self.user.profile.team.pk)
+        self.assertNotEqual(old_team.pk, self.user.profile.team.pk)
+        self.assertEqual(
+            Team.objects.filter(team_code=old_team.team_code).exists(), False
+        )
```
Comment (grahamhoyes) [inline hackathon_site/event/test_api.py]:
```suggestion
        self.assertFalse(
            Team.objects.filter(team_code=old_team.team_code).exists()
        )
```

File: hackathon_site/event/api_views.py [hunk 38]
```diff
@@ -66,3 +67,41 @@ def post(self, request, *args, **kwargs):
...
+    def post(self, request, team_code, *args, **kwargs):
+
+        profile = request.user.profile
+        current_team = profile.team
+
+        self.kwargs["lookup_url_kwarg"] = team_code
+        team = self.get_object()
+
+        if team.profiles.count() >= Team.MAX_MEMBERS:
+            raise ValidationError({"detail": "Team is full."})
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: Minor for consistency:

```suggestion
            raise ValidationError({"detail": "Team is full"})
```

**Review by grahamhoyes — COMMENTED** [review]
Much better!

File: hackathon_site/event/api_views.py [hunk 39]
```diff
@@ -66,3 +66,38 @@ def post(self, request, *args, **kwargs):
...
+
+
+class JoinTeamView(generics.GenericAPIView, mixins.RetrieveModelMixin):
+    permission_classes = [UserHasProfile]
+    serializer_class = TeamSerializer
+    lookup_field = "team_code"
+    queryset = EventTeam.objects.all()
+
+    @transaction.atomic
+    def post(self, request, team_code, *args, **kwargs):
```
Comment (grahamhoyes) [inline hackathon_site/event/api_views.py]: Can actually remove `team_code` from here since it's unused (consumed implicitly by `get_object()` now)

File: hackathon_site/event/views.py [hunk 40]
```diff
@@ -7,15 +7,20 @@
 from django.urls import reverse_lazy
 from django.views.generic.base import TemplateView
 from django.views.generic.edit import FormView
+from django.conf import settings
```
Comment (grahamhoyes) [inline hackathon_site/event/views.py:9]: Take a look at this file again, lots of unused or duplicated imports (I'm guessing from starting working in this file before switching to api_views.py)

File: hackathon_site/hackathon_site/tests.py (hunk 31 above)
Comment (grahamhoyes) [reply hackathon_site/hackathon_site/tests.py]: Still have an unnecessary `,1` at the end (that's obvious for `range()`)

**Review by grahamhoyes — COMMENTED** [review]
Just two minor things, then looks good!

**Review by grahamhoyes — APPROVED** [review]
